README.md
# mccache | github.com/abbeymart/mccache-deno

- Simple and Hash In-Memory Information Cache
- Simple Cache: key-value pair storage
- Hash Cache: hash-hash-value storage
- See the test folder for test cases / scenarios and usage

## API - Simple Cache

- Simple Cache is key-value caching for number, string and object values
- It exposes the following functions: setCache, getCache, deleteCache and
  clearCache

## API - Hash Cache

- Hash Cache: hash-key-value storage for caching number, string and object
  values
- It exposes the following functions: setHashCache, getHashCache,
  deleteHashCache and clearHashCache

## Usage

```ts
// simple-cache
import {
  deleteCache,
  getCache,
  setCache,
} from "https://deno.land/x/mccache@v0.3.1/mod.ts";

// hash-cache
import {
  deleteHashCache,
  getHashCache,
  setHashCache,
} from "https://deno.land/x/mccache@v0.3.1/mod.ts";

// See the test folder for different test cases / scenarios and usage
```


HISTORY.md
# Historical logs

## Version 0.3.1

- Release date: Nov 27, 2023
- Applied-ValueType-from-mcresponse-for-consistency-v0.3.1

## Version 0.3.0

- Release date: Nov 25, 2023
- refactored-value-type-to-generic-v0.3.0

## Version 0.2.2

- Release date: Nov 19, 2022
- Updated-module-version-ref-v0.2.2

## Version 0.2.1

- Release date: Nov 19, 2022
- Updated-module-documentation-v0.2.1

## Version 0.2.0

- Release date: Nov 19, 2022
- Activated-for-deno-module-release-v0.2.0

## Version 0.1.0

- Release date: Nov 19, 2022 | refactored to deno module-v0.1.0
- Simple and Hash In-Memory Cache
- See the test folder for different test cases / scenarios and usage


test.ts
import "./test/simpleCache.test.ts";
import "./test/hashCache.test.ts";


test/hashCache.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-14 | Refactored to deno-module: 2022-11-19
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc: hashCache testing
 */

import {
  assertEquals,
  delay,
  mcTest,
  ObjectType,
  postTestResult,
} from "../test_deps.ts";

import {
  clearHashCache,
  deleteHashCache,
  getHashCache,
  HashCacheParamsType,
  QueryHashCacheParamsType,
  setHashCache,
} from "../src/index.ts";

// test data
const cacheValue: ObjectType = {
    firstName: "Abi",
    lastName: "Akindele",
    location: "Toronto-Canada",
  },
  cacheKey = JSON.stringify({ name: "Tab1", location: "Toronto" }),
  expiryTime = 5, // in seconds
  hashKey = JSON.stringify({ hash1: "Hash1", hash2: "Hash2" });

(async () => {
  console.log("\nHASH-CACHE-TESTING-START\n");
  await mcTest({
    name: "should set and return valid cacheValue",
    testFunc: () => {
      const cacheParams: HashCacheParamsType<ObjectType> = {
        key: cacheKey,
        hash: hashKey,
        value: cacheValue,
        expire: expiryTime,
      };
      const cacheRes = setHashCache(cacheParams);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.value as ObjectType, cacheValue);
        assertEquals(cacheRes.message, "task completed successfully");
        // get cache info
        const getCacheParams: QueryHashCacheParamsType = {
          key: cacheKey,
          hash: hashKey,
        };
        const res = getHashCache(getCacheParams);
        assertEquals(res.ok, true);
        assertEquals(res.value as ObjectType, cacheValue);
        assertEquals(res.message, "task completed successfully");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name: "should clear the cache and return nil/empty value",
    testFunc: () => {
      const cacheRes = clearHashCache();
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.message, "task completed successfully");
        // get cache info
        const getCacheParams: QueryHashCacheParamsType = {
          key: cacheKey,
          hash: hashKey,
        };
        const res = getHashCache(getCacheParams);
        assertEquals(res.ok, false);
        assertEquals(res.message, "cache info does not exist");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name:
      "should set and return valid cacheValue -> before timeout/expiration)",
    testFunc: () => {
      // change the expiry time to 2 seconds
      const cacheParams: HashCacheParamsType<ObjectType> = {
        key: cacheKey,
        hash: hashKey,
        value: cacheValue,
        expire: 2,
      };
      const cacheRes = setHashCache(cacheParams);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.value as ObjectType, cacheValue);
        assertEquals(cacheRes.message, "task completed successfully");
        const getCacheParams: QueryHashCacheParamsType = {
          key: cacheKey,
          hash: hashKey,
        };
        const res = getHashCache(getCacheParams);
        assertEquals(res.ok, true);
        assertEquals(res.value as ObjectType, cacheValue);
        assertEquals(res.message, "task completed successfully");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name: "should return nil/undefined value after timeout/expiration",
    testFunc: async () => {
      await delay(3000);
      const getCacheParams: QueryHashCacheParamsType = {
        key: cacheKey,
        hash: hashKey,
      };
      const res = getHashCache(getCacheParams);
      assertEquals(res.ok, false);
      assertEquals(!res.value, true );
      assertEquals(res.message, "cache expired and deleted");
    },
  });

  await mcTest({
    name:
      "should set and return valid cacheValue (repeat prior to deleteCache testing",
    testFunc: () => {
      // change the expiry time to 10 seconds
      const cacheParams: HashCacheParamsType<ObjectType> = {
        key: cacheKey,
        hash: hashKey,
        value: cacheValue,
        expire: 10,
      };
      const cacheRes = setHashCache(cacheParams);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.value as ObjectType, cacheValue);
        assertEquals(cacheRes.message, "task completed successfully");
        const getCacheParams: QueryHashCacheParamsType = {
          key: cacheKey,
          hash: hashKey,
        };
        const res = getHashCache(getCacheParams);
        assertEquals(res.ok, true);
        assertEquals(res.value as ObjectType, cacheValue);
        assertEquals(res.message, "task completed successfully");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name: "should delete the cache and return nil/undefined value",
    testFunc: () => {
      const delCacheParams: QueryHashCacheParamsType = {
        key: cacheKey,
        hash: hashKey,
      };
      const cacheRes = deleteHashCache(delCacheParams);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.message, "task completed successfully");
        const getCacheParams: QueryHashCacheParamsType = {
          key: cacheKey,
          hash: hashKey,
        };
        const res = getHashCache(getCacheParams);
        assertEquals(res.ok, false);
        assertEquals(!res.value, true );
        assertEquals(res.message, "cache info does not exist");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  postTestResult();
  console.log("\nHASH-CACHE-TESTING-END\n");
})();


test/simpleCache.test.ts
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2020-07-14 | Refactored to deno-module: 2022-11-19
 * @Company: Copyright 2020 Abi Akindele  | mConnect.biz
 * @License: All Rights Reserved | LICENSE.md
 * @Description: mc: simpleCache testing
 */

import { assertEquals, delay, mcTest, postTestResult } from "../test_deps.ts";

import {
  CacheParamsType, CacheResponseType,
  clearCache,
  deleteCache,
  getCache,
  ObjectType,
  setCache,
} from "../src/index.ts";

// test data
const cacheValue: ObjectType = {
    firstName: "Abi",
    lastName: "Akindele",
    location: "Toronto-Canada",
  },
  cacheKey = JSON.stringify({ name: "Tab1", location: "Toronto" }),
  expiryTime = 5; // in seconds

// delay testing (async)

(async () => {
  console.log("\nSIMPLE-CACHE-TESTING-START\n");
  await mcTest({
    name: "should set and return valid cacheValue",
    testFunc: () => {
      const cacheParams: CacheParamsType<ObjectType> = {
        key: cacheKey,
        value: cacheValue,
        expire: expiryTime,
      };
      const cacheRes = setCache(cacheParams);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.value as ObjectType, cacheValue);
        assertEquals(cacheRes.message, "task completed successfully");
        // get cache info
        const res = getCache(cacheKey);
        assertEquals(res.ok, true);
        assertEquals(res.value as ObjectType, cacheValue);
        assertEquals(res.message, "task completed successfully");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name: "should clear the cache and return nil/empty value",
    testFunc: () => {
      const cacheRes = clearCache();
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.message, "task completed successfully");
        // get cache info
        const res = getCache(cacheKey);
        assertEquals(res.ok, false);
        assertEquals(res.message, "cache info does not exist");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name:
      "should set and return valid cacheValue -> before timeout/expiration)",
    testFunc: () => {
      // change the expiry time to 2 seconds
      const cacheParams: CacheParamsType<ObjectType> = {
        key: cacheKey,
        value: cacheValue,
        expire: 2,
      };
      const cacheRes = setCache(cacheParams);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.value as ObjectType, cacheValue);
        assertEquals(cacheRes.message, "task completed successfully");
        const res = getCache(cacheKey);
        assertEquals(res.ok, true);
        assertEquals(res.value as ObjectType, cacheValue);
        assertEquals(res.message, "task completed successfully");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name: "should return nil/undefined value after timeout/expiration",
    testFunc: async () => {
      await delay(3000);
      const res = getCache(cacheKey);
      assertEquals(res.ok, false);
      assertEquals(!res.value, true );
      assertEquals(res.message, "cache expired and deleted");
    },
  });

  await mcTest({
    name:
      "should set and return valid cacheValue (repeat prior to deleteCache testing",
    testFunc: () => {
      // change the expiry time to 10 seconds
      const cacheParams: CacheParamsType<ObjectType> = {
        key: cacheKey,
        value: cacheValue,
        expire: 10,
      };
      const cacheRes = setCache(cacheParams);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.value as ObjectType, cacheValue);
        assertEquals(cacheRes.message, "task completed successfully");
        const res: CacheResponseType<ObjectType> = getCache(cacheKey);
        assertEquals(res.ok, true);
        assertEquals(res.value as ObjectType, cacheValue);
        assertEquals(res.message, "task completed successfully");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  await mcTest({
    name: "should delete the cache and return nil/undefined value",
    testFunc: () => {
      const cacheRes = deleteCache(cacheKey);
      if (cacheRes.ok) {
        assertEquals(cacheRes.ok, true);
        assertEquals(cacheRes.message, "task completed successfully");
        const res = getCache(cacheKey);
        assertEquals(res.ok, false);
        assertEquals(!res.value, true );
        assertEquals(res.message, "cache info does not exist");
      } else {
        assertEquals(cacheRes.ok, false);
      }
    },
  });

  postTestResult();
  console.log("\nSIMPLE-CACHE-TESTING-END\n");
})();


src/types.ts
import { ValueType } from "../deps.ts";

export interface ObjectType {
    [key: string]: ValueType;
}

export type CacheKeyType = string | Record<string, unknown> | number | ObjectType

export interface CacheParamsType<T extends ValueType> {
    key: CacheKeyType;
    value: T;
    expire?: number;
}

export interface HashCacheParamsType<T extends ValueType> {
    key: CacheKeyType;
    hash: CacheKeyType;
    value: T;
    expire?: number;
}

export type ByType = "hash" | "key"

export interface QueryHashCacheParamsType {
    key: CacheKeyType;
    hash: CacheKeyType;
    by?: ByType;
}

export interface CacheValueType<T extends ValueType> {
    value: T;
    expire: number;
}

export interface CacheResponseType<T extends ValueType> {
    ok: boolean;
    message: string;
    value?: T;
}

export type HashCacheValueType<T extends ValueType> = Map<string, CacheValueType<T>>;


src/simpleCache.ts
import {
    CacheKeyType,
    CacheParamsType,
    CacheResponseType,
    CacheValueType,
} from "./types.ts";
import type { ValueType } from "../deps.ts";

// Initialise cache object/dictionary (map)
// deno-lint-ignore no-explicit-any
const mcCache = new Map<string, CacheValueType<any>>();

// secret keyCode for added security
const keyCode = "mcconnect_20200320";

export function setCache<T extends ValueType>(cacheParams: CacheParamsType<T>): CacheResponseType<T | ValueType> {
    // key and value: key:string, value:ValueType, expire:time(seconds)
    const key = cacheParams.key || null;
    const value = cacheParams.value;
    const expire = cacheParams.expire || 300;
    try {
        if (!key || !value) {
            return {
                ok     : false,
                message: "cache key and value are required",
            }
        }
        const cacheKey = JSON.stringify(key) + keyCode;
        mcCache.set(cacheKey, {value: value, expire: Date.now() + expire * 1000});
        return {
            ok     : true,
            message: "task completed successfully",
            value  : mcCache.get(cacheKey)?.value,
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error creating/setting cache information",
        }
    }
}

export function getCache<T extends ValueType>(key: CacheKeyType): CacheResponseType<T> {
    try {
        if (!key) {
            return {
                ok     : false,
                message: "cache key is required",
            }
        }
        const cacheKey = JSON.stringify(key) + keyCode;
        if (mcCache.has(cacheKey)) {
            const cValue = mcCache.get(cacheKey);
            if (cValue && cValue.expire && cValue.expire > Date.now()) {
                return {
                    ok     : true,
                    message: "task completed successfully",
                    value  : cValue.value,
                }
            } else {
                // delete expired cache
                mcCache.delete(cacheKey);
                return {
                    ok     : false,
                    message: "cache expired and deleted",
                }
            }
        } else {
            return {
                ok     : false,
                message: "cache info does not exist",
            }
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error fetching cache information",
        }
    }
}

export function deleteCache<T extends ValueType>(key: CacheKeyType): CacheResponseType<T> {
    try {
        if (!key) {
            return {
                ok     : false,
                message: "cache key is required",
            }
        }
        const cacheKey = JSON.stringify(key) + keyCode;
        if (mcCache.has(cacheKey)) {
            mcCache.delete(cacheKey);
            return {
                ok     : true,
                message: "task completed successfully",
            }
        } else {
            return {
                ok     : false,
                message: "task not completed, cache-key not found",
            }
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error deleting cache information",
        }
    }
}

export function clearCache<T extends ValueType>(): CacheResponseType<T> {
    try {
        // clear the cache object/dictionary (map)
        mcCache.clear();
        return {
            ok     : true,
            message: "task completed successfully",
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error clearing cache",
        }
    }
}



src/hashCache.ts
import {
    CacheResponseType,
    CacheValueType,
    HashCacheParamsType,
    HashCacheValueType,
    QueryHashCacheParamsType,
} from "./types.ts";
import type { ValueType } from "../deps.ts";

// Initialise cache object/dictionary (map)
// Initialise cache object/dictionary (map)
// deno-lint-ignore no-explicit-any
const mcCache = new Map<string, HashCacheValueType<any>>();

// secret keyCode for added security
const keyCode = "mcconnect_20200320";

export function setHashCache<T extends ValueType>(cacheParams: HashCacheParamsType<T>): CacheResponseType<T> {
    // key and value: key:string, hash: string, value:Value, expire:time(seconds)
    const key = cacheParams.key || null;
    const hash = cacheParams.hash || null;
    const value = cacheParams.value;
    const expire = cacheParams.expire || 300;
    try {
        if (!key || !hash || !value) {
            return {
                ok     : false,
                message: "cache key, hash and value are required",
            }
        }
        const cacheKey = JSON.stringify(key) + keyCode;
        const hashKey = JSON.stringify(hash) + keyCode;

        if (!mcCache.has(hashKey)) {
            mcCache.set(hashKey, new Map<string, CacheValueType<T>>());
        }
        const hashValue: CacheValueType<T> = {value: value, expire: Date.now() + expire * 1000};
        mcCache.get(hashKey)?.set(cacheKey, hashValue);
        return {
            ok     : true,
            message: "task completed successfully",
            value  : mcCache.get(hashKey)?.get(cacheKey)?.value,
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error creating/setting cache information",
        }
    }
}

export function getHashCache<T extends ValueType>(cacheParams: QueryHashCacheParamsType): CacheResponseType<T> {
    const key = cacheParams.key;
    const hash = cacheParams.hash;
    try {
        if (!key || !hash) {
            return {
                ok     : false,
                message: "key and hash-key are required",
            }
        }
        const cacheKey = JSON.stringify(key) + keyCode;
        const hashKey = JSON.stringify(hash) + keyCode;
        // get active (non-expired) cache content
        if (mcCache.has(hashKey) && mcCache.get(hashKey)?.has(cacheKey)) {
            const cValue = mcCache.get(hashKey)?.get(cacheKey);
            if (cValue && cValue.expire && cValue.expire > Date.now()) {
                return {
                    ok     : true,
                    message: "task completed successfully",
                    value  : cValue.value,
                }
            } else {
                // delete expired cache
                mcCache.get(hashKey)?.delete((cacheKey));
                return {
                    ok     : false,
                    message: "cache expired and deleted",
                }
            }
        } else {
            return {
                ok     : false,
                message: "cache info does not exist",
            }
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error fetching cache information",
        }
    }
}

export function deleteHashCache<T extends ValueType>(cacheParams: QueryHashCacheParamsType): CacheResponseType<T> {
    const key = cacheParams.key;
    const hash = cacheParams.hash;
    const by = cacheParams.by || "key"
    try {
        if ((!key || !hash) && by === "key") {
            return {
                ok     : false,
                message: "key and hash-key are required",
            }
        }
        const cacheKey = JSON.stringify(key) + keyCode;
        const hashKey = JSON.stringify(hash) + keyCode;
        if (hash && by === "hash" && mcCache.has(hashKey)) {
            mcCache.delete(hashKey);
            return {
                ok     : true,
                message: "task completed successfully",
            }
            // key != "" and hash != "" and by == "hash" and mcCache.hasKey(cacheKey) and mcCache[cacheKey].hasKey(hashKey)
        } else if (by === "key" && mcCache.get(hashKey)?.has(cacheKey)) {
            mcCache.get(hashKey)?.delete(cacheKey);
            return {
                ok     : true,
                message: "task completed successfully",
            }
        } else {
            return {
                ok     : false,
                message: "task could not be completed due to incomplete inputs",
            }
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error deleting cache information",
        }
    }
}

export function clearHashCache<T extends ValueType>(): CacheResponseType<T> {
    try {
        // clear the cache object/dictionary (map)
        mcCache.clear();
        return {
            ok     : true,
            message: "task completed successfully",
        }
    } catch (e) {
        return {
            ok     : false,
            message: e.message ? e.message : "error clearing the cache",
        }
    }
}
